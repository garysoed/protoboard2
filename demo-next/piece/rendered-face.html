<style>
  #summary {
    height: 4rem;
    width: 4rem;
  }
</style>
<pb-lens id="lens">
  <div id="summary">
    <mk-icon id="icon" fit-to="height"></mk-icon>
  </div>
</pb-lens>