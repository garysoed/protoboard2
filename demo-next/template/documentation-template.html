<style>
  #content::slotted(*) {
    padding-left: 1rem;
    padding-right: 1rem;
  }

  #title {
    padding-left: 1rem;
    padding-right: 1rem;
  }
</style>

<section flex>
  <h3 id="title" mk-display></h3>
  <slot id="content" name="content"></slot>
</section>
