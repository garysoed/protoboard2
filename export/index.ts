export {Lens, $lens} from '../src/core/lens';
export {LensDisplay} from '../src/core/lens-display';

// action
export {pickAction} from '../src/action/pick-action';

// coordinate
export {Cartesian} from '../src/coordinate/cartesian';
export {Coordinate} from '../src/coordinate/coordinate';
export {Indexed} from '../src/coordinate/indexed';

// region
export {Deck, $deck, DeckSpec} from '../src/region/deck';
export {Slot, $slot, SlotSpec, slotSpec} from '../src/region/slot';

// core
export {$createSpecEntries, $createSpecMap} from '../src/objects/object-create-spec';
export {ContentSpec} from '../src/payload/is-container';
export {ObjectSpec} from '../src/types/object-spec';
export {$render, Render} from '../src/objects/render';
export {$active, Active, activeSpec} from '../src/core/active';
export {$$activeSpec} from '../src/objects/active-spec';

// piece
export {$d1, D1, D1Spec, d1Spec} from '../src/piece/d1';
export {$d2, D2, D2Spec} from '../src/piece/d2';
export {$d6, D6, D6Spec} from '../src/piece/d6';

// tiling
export {rectOrthogonal, RectOrthogonal, Direction as RectOrthogonalDirection, distance as rectOrthogonalDistance} from '../src/tiling/rect-orthogonal';
export {vhex, VHex, Direction as VHexDirection, distance as vhexDistance} from '../src/tiling/vhex';
export {Tile} from '../src/tiling/types';
