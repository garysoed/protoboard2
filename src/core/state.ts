import { Observable } from 'rxjs';

/**
 * State of a game during its runtime.
 *
 * @thModule core
 */
export interface State {
  /**
   * Identifies the object. Must be globally unique.
   */
  readonly id: string;

  /**
   * Used to map to a function that generates the object.
   */
  readonly type: string;

  /**
   * Used to generate the object.
   */
  readonly payload: ReadonlyMap<string, Observable<unknown>>;
}
