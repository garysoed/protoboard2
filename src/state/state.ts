import { BehaviorSubject } from 'rxjs';

/**
 * State of a game during its runtime.
 *
 * @thModule core
 */
export interface State<P extends object> {
  /**
   * Identifies the object. Must be globally unique.
   */
  readonly id: string;

  /**
   * Used to map to a function that generates the object.
   */
  readonly type: string;

  /**
   * Used to generate the object.
   * TODO: Instead of BehaviorSubject, this should be split into Observable and Subject/
   */
  readonly payload: {readonly [K in keyof P]: BehaviorSubject<P[K]>};
}
