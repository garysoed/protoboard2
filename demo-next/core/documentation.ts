import {cache} from 'gs-tools/export/data';
import {renderTheme} from 'mask';
import {Context, Ctrl, DIV, ocase, query, registerCustomElement, renderElement, RenderSpec} from 'persona';
import {Observable} from 'rxjs';
import {map} from 'rxjs/operators';

import {D1_DEMO} from '../piece/d1';
import {D2_DEMO} from '../piece/d2';
import {D6_DEMO} from '../piece/d6';

import template from './documentation.html';
import {INSTRUCTION} from './instruction';
import {$locationService} from './location-service';
import {getPageSpec, PageSpec} from './page-spec';


const $documentation = {
  shadow: {
    root: query('#root', DIV, {
      content: ocase<PageSpec|null>('#content'),
    }),
  },
};

export class Documentation implements Ctrl {
  constructor(private readonly $: Context<typeof $documentation>) { }

  @cache()
  get runs(): ReadonlyArray<Observable<unknown>> {
    return [
      renderTheme(this.$),
      $locationService.get(this.$.vine).location$.pipe(map(location => getPageSpec(location.type))).pipe(
          this.$.shadow.root.content(spec => this.renderContent(spec)),
      ),
    ];
  }

  @cache()
  private renderContent(spec: PageSpec|null): RenderSpec|null {
    if (!spec) {
      return null;
    }

    return renderElement({spec: {}, registration: spec.registration});
  }
}

export const DOCUMENTATION = registerCustomElement({
  ctrl: Documentation,
  deps: [
    // CanvasDemo,
    D1_DEMO,
    D2_DEMO,
    D6_DEMO,
    // DeckDemo,
    INSTRUCTION,
    // SlotDemo,
  ],
  spec: $documentation,
  tag: 'pbd-documentation',
  template,
});