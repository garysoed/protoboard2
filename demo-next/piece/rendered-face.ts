import {cache} from 'gs-tools/export/data';
import {ICON, registerSvg, renderTheme} from 'mask';
import {Context, Ctrl, H3, iattr, id, otext, P, registerCustomElement} from 'persona';
import {Observable} from 'rxjs';
import {map} from 'rxjs/operators';

import cardBackSvg from '../asset/card_back.svg';
import cardFrontSvg from '../asset/card_front.svg';
import coinFrontSvg from '../asset/coin.svg';
import coinBackSvg from '../asset/coin_back.svg';
import dicePip1Svg from '../asset/dice_pip_1.svg';
import dicePip2Svg from '../asset/dice_pip_2.svg';
import dicePip3Svg from '../asset/dice_pip_3.svg';
import dicePip4Svg from '../asset/dice_pip_4.svg';
import dicePip5Svg from '../asset/dice_pip_5.svg';
import dicePip6Svg from '../asset/dice_pip_6.svg';
import gemSvg from '../asset/gem.svg';
import meepleSvg from '../asset/meeple.svg';

import template from './rendered-face.html';


export enum FaceType {
  CARD_BACK = 'cardback',
  CARD_FRONT = 'cardfront',
  COIN_BACK = 'coinback',
  COIN_FRONT = 'coinfront',
  DICE_PIP_1 = 'dicepip1',
  DICE_PIP_2 = 'dicepip2',
  DICE_PIP_3 = 'dicepip3',
  DICE_PIP_4 = 'dicepip4',
  DICE_PIP_5 = 'dicepip5',
  DICE_PIP_6 = 'dicepip6',
  GEM = 'gem',
  MEEPLE = 'meeple',
}

const $renderedFace = {
  host: {
    faceType: iattr('face-type'),
  },
  shadow: {
    icon: id('icon', ICON),
    name: id('name', H3, {
      text: otext(),
    }),
    details: id('details', P),
  },
};


class RenderedFace implements Ctrl {
  constructor(private readonly $: Context<typeof $renderedFace>) { }

  get runs(): ReadonlyArray<Observable<unknown>> {
    return [
      renderTheme(this.$),
      this.icon$.pipe(this.$.shadow.icon.icon()),
      this.faceName$.pipe(this.$.shadow.name.text()),
    ];
  }

  @cache()
  private get icon$(): Observable<string> {
    return this.$.host.faceType.pipe(map(faceType => faceType ?? ''));
  }

  @cache()
  private get faceName$(): Observable<string> {
    return this.$.host.faceType.pipe(
        map(faceType => {
          switch (faceType) {
            case FaceType.CARD_BACK:
              return 'Card Back';
            case FaceType.CARD_FRONT:
              return 'Card Front';
            case FaceType.COIN_BACK:
              return 'Coin Back';
            case FaceType.COIN_FRONT:
              return 'Coin Front';
            case FaceType.DICE_PIP_1:
              return 'Dice 1 (pip)';
            case FaceType.DICE_PIP_2:
              return 'Dice 2 (pip)';
            case FaceType.DICE_PIP_3:
              return 'Dice 3 (pip)';
            case FaceType.DICE_PIP_4:
              return 'Dice 4 (pip)';
            case FaceType.DICE_PIP_5:
              return 'Dice 5 (pip)';
            case FaceType.DICE_PIP_6:
              return 'Dice 6 (pip)';
            case FaceType.GEM:
              return 'Gem';
            case FaceType.MEEPLE:
              return 'Meeple';
            default:
              return '';
          }
        }),
    );
  }
}

export const RENDERED_FACE = registerCustomElement({
  configure: vine => {
    registerSvg(vine, FaceType.CARD_BACK, {type: 'embed', content: cardBackSvg});
    registerSvg(vine, FaceType.CARD_FRONT, {type: 'embed', content: cardFrontSvg});
    registerSvg(vine, FaceType.COIN_BACK, {type: 'embed', content: coinBackSvg});
    registerSvg(vine, FaceType.COIN_FRONT, {type: 'embed', content: coinFrontSvg});
    registerSvg(vine, FaceType.GEM, {type: 'embed', content: gemSvg});
    registerSvg(vine, FaceType.MEEPLE, {type: 'embed', content: meepleSvg});
    registerSvg(vine, FaceType.DICE_PIP_1, {type: 'embed', content: dicePip1Svg});
    registerSvg(vine, FaceType.DICE_PIP_2, {type: 'embed', content: dicePip2Svg});
    registerSvg(vine, FaceType.DICE_PIP_3, {type: 'embed', content: dicePip3Svg});
    registerSvg(vine, FaceType.DICE_PIP_4, {type: 'embed', content: dicePip4Svg});
    registerSvg(vine, FaceType.DICE_PIP_5, {type: 'embed', content: dicePip5Svg});
    registerSvg(vine, FaceType.DICE_PIP_6, {type: 'embed', content: dicePip6Svg});
  },
  ctrl: RenderedFace,
  deps: [
    ICON,
  ],
  spec: $renderedFace,
  tag: 'pbd-rendered-face',
  template,
});