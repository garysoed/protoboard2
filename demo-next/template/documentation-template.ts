import {cache} from 'gs-tools/export/data';
import {mapNullableTo} from 'gs-tools/export/rxjs';
import {renderTheme} from 'mask';
import {Context, Ctrl, H3, iattr, id, otext, registerCustomElement} from 'persona';
import {Observable} from 'rxjs';

import template from './documentation-template.html';


const $documentationTemplate = {
  host: {
    label: iattr('label'),
  },
  shadow: {
    title: id('title', H3, {
      text: otext(),
    }),
  },
};

class DocumentationTemplate implements Ctrl {
  constructor(private readonly $: Context<typeof $documentationTemplate>) { }

  @cache()
  get runs(): ReadonlyArray<Observable<unknown>> {
    return [
      renderTheme(this.$),
      this.$.host.label.pipe(mapNullableTo(''), this.$.shadow.title.text()),
    ];
  }
}

export const DOCUMENTATION_TEMPLATE = registerCustomElement({
  ctrl: DocumentationTemplate,
  spec: $documentationTemplate,
  tag: 'pbd-documentation-template',
  template,
});
