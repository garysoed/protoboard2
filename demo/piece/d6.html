<pbd-piece-template id="template" label="D6" face-count="6">
  <p mk-body-1 slot="details">
    A D6 is the most is a piece with 6 faces. Some examples include dice.
  </p>
</pbd-piece-template>
