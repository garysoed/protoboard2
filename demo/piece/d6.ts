import {immutablePathSource, rootStateIdSource} from 'grapevine';
import {mutableState} from 'gs-tools/export/state';
import {BaseThemedCtrl, Icon, _p} from 'mask';
import {element, PersonaContext} from 'persona';
import {Observable, of} from 'rxjs';

import {$registerRenderObject, $slot, D6, d6Spec, D6Spec, Lens, Slot, slotSpec, SlotSpec} from '../../export';
import {FaceType, RenderedFace} from '../core/rendered-face';
import {DocumentationTemplate} from '../template/documentation-template';

import template from './d6.html';
import {renderPiece} from './render-piece';


interface State {
  readonly diceSlot: SlotSpec;
}

const $diceId = rootStateIdSource<D6Spec>(() => d6Spec({}));
const $dicePath = immutablePathSource($diceId);

const $stateId = rootStateIdSource<State>(vine => ({
  diceSlot: slotSpec({contentsId: mutableState([$dicePath.get(vine)])}),
}));
const $diceSlotPath = immutablePathSource($stateId, state => state._('diceSlot'));

export const $d6Demo = {
  tag: 'pbd-d6',
  api: {},
};

const $ = {
  diceSlot: element('diceSlot', $slot, {}),
};

@_p.customElement({
  ...$d6Demo,
  configure: vine => {
    const registerRenderObject = $registerRenderObject.get(vine);
    const dicePath = $dicePath.get(vine);
    registerRenderObject(
        dicePath,
        renderPiece(
            [
              FaceType.DICE_PIP_1,
              FaceType.DICE_PIP_2,
              FaceType.DICE_PIP_3,
              FaceType.DICE_PIP_4,
              FaceType.DICE_PIP_4,
              FaceType.DICE_PIP_5,
            ],
            dicePath,
        ),
    );
  },
  dependencies: [
    D6,
    DocumentationTemplate,
    Icon,
    Lens,
    RenderedFace,
    Slot,
  ],
  template,
})
export class D6Demo extends BaseThemedCtrl<typeof $> {
  constructor(context: PersonaContext) {
    super(context, $);
  }

  protected get renders(): ReadonlyArray<Observable<unknown>> {
    return [
      this.renderers.diceSlot.objectPath(of($diceSlotPath.get(this.context.vine))),
    ];
  }
}
