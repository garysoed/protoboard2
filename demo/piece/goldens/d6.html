<pbd-d6><style>
  .slotContainer {
    display: inline-block;
    position: relative;
    width: 12rem;
    height: 12rem;
  }
</style>

<pbd-documentation-template mk-body-1="" label="D6"><style>
  #content::slotted(*) {
    padding-left: 1rem;
    padding-right: 1rem;
  }

  #title {
    padding-left: 1rem;
    padding-right: 1rem;
  }
</style>

<section flex="">
  <h3 id="title" mk-display="">D6</h3>
  <slot id="content" name="content"><div slot="content">
    <p>
      A D6 is a piece with six faces. The most obvious example is a six sided die.
    </p>
    <p>
      </p><div class="slotContainer">
        <pb-slot id="diceSlot" object-path="i::4"><style>
  #root {
    align-items: center;
    justify-content: center;
    background-color: var(--mkThemeBG);
    border: 1px solid var(--mkThemeFG);
    box-sizing: border-box;
    flex-wrap: wrap;
  }
</style>

<div id="root" fill-parent="" mk-action-2="" layout="row">
  <!-- #content--><pb-d6 object-path="i::2"><slot id="face" name="face-0"><pbd-rendered-face face-type="dicepip1" slot="face-0"><style>
  #summary {
    height: 4rem;
    width: 4rem;
  }
</style>
<pb-lens><style>
  #details {
    display: none;
  }
</style>

<div id="root">
  <slot name="summary"><div id="summary" slot="summary">
    <mk-icon id="icon" aria-hidden="true" role="presentation" icon="dicepip1"><style>
  :host {
    pointer-events: none;
    user-select: none;
  }

  #root > svg #foreground {
    fill: var(--mkThemeFG) !important;
  }
</style>
<span id="root">
  <!-- #content --><svg xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:cc="http://creativecommons.org/ns#" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:svg="http://www.w3.org/2000/svg" xmlns="http://www.w3.org/2000/svg" xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd" xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape" height="100%" fill="#000000" version="1.1" x="0px" y="0px" viewBox="0 0 24 24" enable-background="new 0 0 100 100" xml:space="preserve" id="svg4170" inkscape:version="1.0 (4035a4f, 2020-05-01)" sodipodi:docname="dice_pip_1.svg"><metadata id="metadata4178"><rdf:RDF><cc:Work rdf:about=""><dc:format>image/svg+xml</dc:format><dc:type rdf:resource="http://purl.org/dc/dcmitype/StillImage"></dc:type><dc:title></dc:title></cc:Work></rdf:RDF></metadata><defs id="defs4176"></defs><sodipodi:namedview pagecolor="#ffffff" bordercolor="#666666" borderopacity="1" objecttolerance="10" gridtolerance="10" guidetolerance="10" inkscape:pageopacity="0" inkscape:pageshadow="2" inkscape:window-width="1393" inkscape:window-height="825" id="namedview4174" showgrid="false" inkscape:zoom="11.538666" inkscape:cx="7.1434192" inkscape:cy="17.107915" inkscape:window-x="0" inkscape:window-y="25" inkscape:window-maximized="0" inkscape:current-layer="svg4170" inkscape:document-rotation="0"></sodipodi:namedview><path d="M 4.8535156 0 C 2.1648164 5.9211895e-16 0 2.1648164 0 4.8535156 L 0 19.146484 C -5.9211895e-16 21.835184 2.1648164 24 4.8535156 24 L 19.146484 24 C 21.835184 24 24 21.835184 24 19.146484 L 24 4.8535156 C 24 2.1648164 21.835184 0 19.146484 0 L 4.8535156 0 z M 11.939453 8 A 4 4 0 0 1 12 8 A 4 4 0 0 1 16 12 A 4 4 0 0 1 12 16 A 4 4 0 0 1 8 12 A 4 4 0 0 1 11.939453 8 z " style="fill:#000000;stroke-width:2.23856" id="foreground"></path></svg>
</span>
<link rel="stylesheet" href="base/test.css"></mk-icon>
  </div></slot>
</div>
<slot id="details" name="details"><div slot="details">
    <h3 id="name">Dice 1 (pip)</h3>
    <p mk-body-1="">More detailed information on the piece goes here.</p>
  </div></slot><link rel="stylesheet" href="base/test.css"></pb-lens><link rel="stylesheet" href="base/test.css"></pbd-rendered-face></slot>
<link rel="stylesheet" href="base/test.css"></pb-d6>
</div>
<link rel="stylesheet" href="base/test.css"></pb-slot>
      </div>
    <p></p>
  </div></slot>
</section>
<link rel="stylesheet" href="base/test.css"></pbd-documentation-template>
<link rel="stylesheet" href="base/test.css"></pbd-d6>